src/coreclr/vm/interpreter.cpp

```diff
@@ -10903,6 +10922,169 @@ void Interpreter::DoGetArrayDataReference()
     OpStackTypeSet(ind, InterpreterType(CORINFO_TYPE_BYREF));
 }
 
+void Interpreter::DoIsReferenceOrContainsReferences(CORINFO_METHOD_HANDLE method)
+{
+    CONTRACTL{
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    } CONTRACTL_END;
+
+    CORINFO_SIG_INFO sigInfoFull;
+    {
+        GCX_PREEMP();
+        m_interpCeeInfo.getMethodSig(method, & sigInfoFull, nullptr);
+    }
+
+    MethodTable* typeArg = GetMethodTableFromClsHnd(sigInfoFull.sigInst.methInst[0]);
+    OpStackSet<BOOL>(m_curStackHt, typeArg->IsByRefLike() || typeArg->ContainsPointers());
+    OpStackTypeSet(m_curStackHt, InterpreterType(CORINFO_TYPE_INT));
+    m_curStackHt++;
+}
+
+bool Interpreter::DoInterlockedCompareExchange()
+{
+    CONTRACTL{
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    } CONTRACTL_END;
+
+    // These CompareExchange are must-expand:
+    // 
+    //  long   CompareExchange(ref long location1, long value, long comparand)
+    //  int    CompareExchange(ref int location1, int value, int comparand)
+    //  ushort CompareExchange(ref ushort location1, ushort value, ushort comparand)
+    //  byte   CompareExchange(ref byte location1, byte value, byte comparand)
+    //
+    // Detect these by comparand's type (stack - 1):
+    unsigned comparandInd = m_curStackHt - 1;
+    unsigned valueInd = m_curStackHt - 2;
+    unsigned locationInd = m_curStackHt - 3;
+    switch (OpStackTypeGet(comparandInd).ToCorInfoType())
+    {
+        case CORINFO_TYPE_LONG:
+            m_curStackHt -= 3;
+            OpStackSet<int64_t>(m_curStackHt, InterlockedCompareExchange64(
+                OpStackGet<int64_t*>(locationInd),
+                OpStackGet<int64_t>(valueInd),
+                OpStackGet<int64_t>(comparandInd)));
+            OpStackTypeSet(m_curStackHt, InterpreterType(CORINFO_TYPE_LONG));
+            m_curStackHt++;
+            return true;
+
+        case CORINFO_TYPE_INT:
```

Comment:
Won't this be the case hit for small types since there is no small IL stack type? Seems like the determination needs to be based on the method signature.

Reply:
good catch!